Mobile updates

svanderburg opened the pull request:
###### Motivation for this change

Replaces the androidenv implementation by a new implementation that uses:
* The latest repository specs to generate expressions from (repository2-1.xml)
* A more robust generation strategy that exposes many more plugins
* A more robust autopatching strategy for ELF binaries
* A function composition strategy that makes it easier to maintain androidenv.buildApp and the corresponding Android SDK features a build may need

Furthermore, it updates other mobile build aspects:
* A refactored xcodeenv implementation (iphone build environment)
* A refactored titaniumenv implementation using the new androidenv and xcodeenv environments

###### Status:

Done:
- [X] New androidenv implementation
- [X] Refactored xcodeenv implementation
- [X] Refactored titaniumenv implementation
- [X] Added docs for androidenv to the manual
- [X] Added docs for xcodeenv to the manual
- [X] Added docs for titaniumenv to the manual

Broken packages ported:
- [X] apktool
- [X] adbfs-rootless
- [X] adb-sync
- [X] flashtool
- [X] scrcpy
- [X] gnirehtet

Still to do:
- [x] Use modified autopatchelfHook that can also be used as a CLI
- [x] Add license accepted flag, because the Android SDK requires you to accept a license agreement
- [ ] Readd androidenv.buildGradleApp
- [x] Fix cross compilation (stdenv.cross) and androidPkgs

###### Things done

<!-- Please check what applies. Note that these are not hard requirements but merely serve as information for reviewers. -->

- [ ] Tested using sandboxing ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS, or option `sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file) on non-NixOS)
- Built on platform(s)
   - [X] NixOS
   - [X] macOS
   - [X] other Linux distributions
- [ ] Tested via one or more NixOS test(s) if existing and applicable for the change (look inside [nixos/tests](https://github.com/NixOS/nixpkgs/blob/master/nixos/tests))
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [X] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Determined the impact on package closure size (by running `nix path-info -S` before and after)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

svanderburg committed:
Replace xcodeenv with new implementation

svanderburg committed:
Replace titaniumenv with new implementation

svanderburg committed:
Port packages with Android dependencies to new androidenv

svanderburg committed:
Add documentation section for Android

svanderburg committed:
Add Titanium documentation section

nixos-discourse commented:
This pull request has been mentioned on **Nix community**. There might be relevant details there:

https://discourse.nixos.org/t/my-revised-experimental-android-build-functions/923/12

svanderburg commented:
@Ericson2314 I could use your help for the stdenv.cross

svanderburg commented:
@matthewbauer I could use your help for readding `androidenv.buildGradleApp`

Mic92 commented on pkgs/development/mobile/androidenv/autopatchelf/autopatchelf:
```diff
@@ -0,0 +1,187 @@
+#!/bin/bash -e
```
cc @aszlig

aszlig replied:
@Mic92: See https://github.com/NixOS/nixpkgs/pull/50802.

svanderburg commented on pkgs/development/mobile/androidenv/ndk-bundle/make_standalone_toolchain.py_18.patch:
```diff
@@ -0,0 +1,72 @@
+                      platforms_path, host_tag):
+     """Create a standalone toolchain."""
+     copy_directory_contents(gcc_path, install_path)
++    os.system('chmod -R +w "{}"'.format(install_path))
```
I just simplified this patch. Originally, I stole the patch from somebody else, but I think it won't harm to get rid of code duplication.

svanderburg commented:
@Mic92 I simplified the patch

svanderburg commented on doc/languages-frameworks/ios.section.md:
```diff
@@ -0,0 +1,219 @@
+---
+title: Android
```
Hmm copy-and-paste error :)

Ericson2314 commented:
@svanderburg so I think the "android ndk packages" file I made you'll basically want to keep as-is. The point of it is to take the NDK stuff, however that is gotten, and chop it up into things that we can replace normal nixpkgs derivations with (gcc, bionic libc, etc). Maybe with your new finer-grained plugin approach, it will be easier to get just those bits rather than downloading more and then throwing array the rest, but most of the complicatedness in that file has to do with the fact that tools and library overrides belong in different bootstrapping stages, and that much remains the same.

svanderburg committed:
Replace androidenv by new implementation

svanderburg committed:
Add documentation for building iOS apps

svanderburg commented:
I just rebased against the latest master. This resolves a conflict and allows me to use the modified autoPatchelf executable implemented in issue #50802

svanderburg commented:
@Ericson2314 Thanks! I'll start investigating a reintegration approach of these package soon.

The stdenv/cross environment is also broken because of the new packaging structure -- we now need to refer to the Android NDK (that is now bundled as a plugin with the Android SDK) in a different way.

svanderburg committed:
Fix inclusion of sources

svanderburg commented:
@Ericson2314 I did a preliminary investigation and there are some practical concerns for which I need to know your opinion.

The `androidndk-pkgs.nix` file is basically a set of packages that you want to cross compile with the Android NDK, right?

I'm not sure if putting this file in `androidenv/` is the most logical place. For example, packages built with OpenJDK or Python are also not technically in the same "environment". The purpose of the androidenv component is to encapsulate the Android SDK manager and its plugins.

Do you mind if I put this expression in `pkgs/development/androidndk-packages` ? This is a location where also packages for other kinds of development environments reside, e.g. `pkgs/development/python-modules/`, `pkgs/development/node-modules/` etc. I think this also makes it easier to implement future refactorings of the `androidenv`

The other problem is the Android NDK package -- the newest Android SDK also bundles the NDK as a plugin that you can install with the SDK manager. This is how I integrate the Android NDK in Nixpkgs.

I noticed that `stdenv/cross` supports various kinds of NDK versions. Unfortunately, the `repository2-1.xml` config file provided by Google (that I use in the new generation process) only provides one NDK version, namely: `18b`

Is it still desired to support older NDK versions? If so, then we must package these older NDK versions differently somehow, by using the XML files of older SDK versions.

svanderburg commented:
@Ericson2314 I now started investigating whether I can separate the androidndkPackages from androidenv (on the upstream master branch).

The idea is that I first try to separate it and check whether the functionality still works (e.g. whether there are no regressions). If this succeeds, then I apply the patches that provide my new SDK implementation. Finally, I will fix the androidndk packages to work with the NDK bundled with the new SDK implementation (version 18b).

Is there simple way to test how this cross compiling stuff works? I noticed that there is `bionic` in `all-packages.nix`, but I can't seem to build it by running:

```bash
$ nix-build -A bionic
```
The above command throws an assertion error.

I probably need to supply a bunch of cross compiling parameters, but from the documentation it is not clear to me how I can, for example, build on my `x86_64-linux` system a `bionic` implementation for ARM or x86_64 (or whatever targets are supported).

Is there simple testcase that I can run to make `bionic` build properly?

Ericson2314 commented:
```
$ nix-build -A pkgsCross.aarch64-android-prebuilt.hello
```
will build GNU hello with ndk's GCC and bionic.

svanderburg committed:
Separate androidndkpkgs from androidenv

svanderburg commented:
@Ericson2314 I just created a new branch (forked from the current Nixpkgs) in which I implemented the separation (https://github.com/svanderburg/nixpkgs/commit/13d6088be845c82802bb70f6f7514805dff15901). Your testcase seems to pass.

What I basically did: I moved androidPkgs from androidenv to the top-level expression and moved all androidPkgs stuff to `pkgs/development/androidndk-pkgs`. 

Next thing is trying to rebase my SDK changes on top of this change and fixing to make it work with the Android NDK 18b.

Ericson2314 commented:
Looks great!

svanderburg committed:
Fix standalone generation script

svanderburg commented:
I just rebased again to include the android NDK pkgs separation.

@Ericson2314 I have managed to partially fix the cross compilation, but I need some help from you. The example package that you provided me now seems to properly evaluate. However, when I run the build, but the configure script still fails to properly detect the compiler. Inspecting `config.log` shows me that it can't find `-lgcc`, which makes sense since the newest Android NDK is clang-based.

I have created a separate WIP branch (https://github.com/svanderburg/nixpkgs/tree/mobile-updates-wip) that contains my changes so far: https://github.com/svanderburg/nixpkgs/commit/63f235c81ef1259c2cd9b1d7d8b9b44e5aac2b76

I think quite a few things in the Android 18b NDK have changed compared to the latest previous version 17c. For example, symlinking to the cross compilation executables no longer seems to work because they are shell script that themselves are wrappers using relative paths.

Moreover, since the newest Android SDK distribution only supports NDK 18b so far, and the older NDK versions are no longer supported, we can probably also remove the code that is specific to these older version, e.g. 17c, 10e, 8e.

Can you have look at it?

Ericson2314 commented:
I wonder if `-lgcc` has something to do with one of our wrapper scripts. `hello` itself ought not to care about those details.

svanderburg committed:
adb-sync: fix to make it work with platform-tools

svanderburg committed:
Remove redundant getBin function

svanderburg committed:
Reintroduce licenseAccepted option

svanderburg committed:
Fix adb program module

svanderburg commented:
I rebased again to fix merge conflicts. Its now nearing completion: cross compiling is still broken and the `androidenv.buildGradleApp {}` function is missing.

svanderburg commented:
@Ericson2314 Let me know if you need any help fixing the stdenv/cross to make Android builds work again. I think is the biggest remaining obstacle to get this patch set integrated.

svanderburg commented:
@ali-abrar @matthewbauer I could still use your help in getting the `androidenv.buildGradleApp {}` function ported. I could probably rewrite it, but I'm not using it myself and I don't have the means to validate it.

svanderburg committed:
Initial attempt to restore Android NDK cross building

svanderburg commented:
@Ericson2314 I think I have managed to make the fixes on my own. Apparently, what I learned is that libgcc.a is still required in a clang build if GNU Binutils are used for linking. I modified the bintools package in such a way that the required libgcc.a can be found.

Another problem I ran into is that the `asm/types.h` header can no longer be found when I try to compile GNU Hello with bionic. This is most likely caused by a slightly different sysroot directory organization in the Android NDK. I modified the bionic expression in such a way that these target specific ASM headers are copied in the right locations.

After making these modifications I have successfully compiled GNU Hello for aarch64 on Android.

Since I'm not using this stuff extensively, let me know if there is anything else that may require fixing. My changes are in revision 5142862

I think everything now is fixed and there are, as far as I can see, no breaking changes so I believe there is nothing preventing me from removing the WIP status of this pull request.

Ericson2314 commented:
Great job! And sorry I didn't get to it before. I am taking a look now. BTW since you are doing xcode stuff, you may also want to checkout out what we did for ios cross here: https://github.com/nixos/nixpkgs/blob/master/pkgs/os-specific/darwin/xcode/default.nix using https://github.com/nixos/nixpkgs/blob/master/pkgs/os-specific/darwin/ios-sdk-pkgs/default.nix for the cross packages.

svanderburg commented on pkgs/development/androidndk-pkgs/androidndk-pkgs.nix:
```diff
@@ -109,18 +118,16 @@ rec {
       cp -r ${includePath} $out/include
+      chmod u+w $out/include
+      cp -r ${asmIncludePath}/* $out/include
       chmod +w $out/include
```
Hmm is not needed. Basically I needed to restore write permissions first before I could copy the asm headers, but I didn't bother to change the original statement that restores write permissions

svanderburg commented on pkgs/development/androidndk-pkgs/androidndk-pkgs.nix:
```diff
@@ -54,8 +56,15 @@ rec {
       for prog in ${ndkBinDir}/${targetInfo.triple}-*; do
         prog_suffix=$(basename $prog | sed 's/${targetInfo.triple}-//')
-        ln -s $prog $out/bin/${stdenv.targetPlatform.config}-$prog_suffix
+        cat > $out/bin/${stdenv.targetPlatform.config}-$prog_suffix <<EOF
```
Yes, symlinking does not work for the executables in Android NDK 18b. The executables we refer to are bash scripts that load executable from relative locations and have the expectation that they are loaded from the Android NDK directory. If we use symlinks, then it will look relative to the path of the symlinks, breaking the executable because these paths are invalid.

svanderburg committed:
Remove redundant chmod

svanderburg commented:
@Ericson2314 I just removed the redundant chmod instruction. I also removed the WIP status from this pull request.

I will give people just a little while to provide their feedback. I think these changes can be integrated soon.

svanderburg committed:
Remove relatedPackages to fix ofborg evaluation

svanderburg commented:
Since I have not received any further comments, I'm planning to integrate this PR tomorrow

7c6f434c commented:
Hm. It looks like `adb` per se has Apache2 source, and previously it was provided as a free package (while the SDK itself was unfree). I am trying to keep unfree stuff in a separately controlled user-owned package set and I want adb in a free and root-owned package set. Is it currently possible?

nixos-discourse commented:
This pull request has been mentioned on **Nix community**. There might be relevant details there:

https://discourse.nixos.org/t/building-openwrt-with-nix-discuss-on-35c3/1777/2

bkchr commented:
@svanderburg Did you ever tried the `ndk-build`? For me it is failing :( It seems to pick up the system `ld.gold` linker. Any ideas on how to fix this?

Artturin commented:
`buildGradleApp` was never readded

nixos-discourse commented:
This pull request has been mentioned on **NixOS Discourse**. There might be relevant details there:

https://discourse.nixos.org/t/using-androidenv-error-functionargs-requires-a-function/32773/5

Artturin commented:
Readding in https://github.com/NixOS/nixpkgs/pull/254082
